crates/cli-support/src/js/rust2js.rs

```diff
@@ -499,6 +499,37 @@ impl<'a, 'b> Rust2Js<'a, 'b> {
         Ok(())
     }
 
+    /// Returns whether this shim won't actually do anything when called other
+    /// than forward the invocation somewhere else.
+    ///
+    /// This is used as an optimization to wire up imports directly where
+    /// possible and avoid a shim in some circumstances.
+    pub fn is_noop(&self) -> bool {
+        let Rust2Js {
+            // fields which may affect whether we do nontrivial work
+            catch,
+            finally,
+            js_arguments,
+            prelude,
+            ret_expr,
+            variadic,
+
+            // all other fields, listed explicitly here so if one is added we'll
+            // trigger a nonexhaustive error.
+            arg_idx: _,
+            cx: _,
+            global_idx: _,
+            shim_arguments: _,
+        } = self;
+
+        !catch &&
+            !variadic &&
+            prelude.is_empty() &&
+            finally.is_empty() &&
+            (ret_expr == "JS;" || ret_expr == "return JS;") &&
+            js_arguments.iter().all(|arg| !arg.contains("(") && !arg.contains(" "))
```

Comment:
These last two lines on this predicate are a bit funky... What is going on here?

Reply:
Heh, good question! I'll add some comments as well, but the gist is that this is trying to detect if nontrivial transformations are needed on the JS return value or any of the arguments, where only bare return expressions are allowed and arguments which are directly forwarding the JS versions (containing `(` typically means a method call and spaces mean something like a ternary expression)